### PR TITLE
Update password endpoint for byron

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#1483 

# Overview

- [x] I have added `ByronWalletPasspphraseData`
- [x] I have added endpoint to Api modules
- [x] I have added endpoint to swagger
- [x] I have updated Api.TypesSpec and Malformed
- [x] I have added integration tests for updating valid paswd->valid passwd
- [x] I have introduced Password "byron-raw"
- [x] I have enabled null passwords in `ByronWalletPostData` 
- [x] I have updated the rest of codebase bacause of it
- [x] I have added integration tests for updating no passwd -> valid passwd 

# Comments

### CODE REVIEW

**Comment by paweljakubas** [comment]
bors try

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 1]
```diff
@@ -469,6 +469,11 @@ data WalletPutPassphraseData = WalletPutPassphraseData
     , newPassphrase :: !(ApiT (Passphrase "raw"))
     } deriving (Eq, Generic, Show)
 
+data ByronWalletPutPassphraseData = ByronWalletPutPassphraseData
+    { oldPassphrase :: !(Maybe (ApiT (Passphrase "raw")))
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]: This won't work unfortunately ... not only does the old passphrase can be empty, but it can also be less than 10 characters (which the new backend doesn't have as a limitation). 

My suggestion would be to have a `ApiT (Passphrase "byron-raw")`, for which we can define a different minimum length that is different from `Passphrase "raw"`, and coerce the types in the API handlers.

File: lib/cli/src/Cardano/CLI.hs [hunk 2]
```diff
@@ -1045,22 +1047,39 @@ cmdWalletUpdatePassphrase mkClient =
...
+instance PassphraseUpdate ByronWalletPutPassphraseData where
+    producePassphraseData (OldNewPassword old new) =
+        if old == Passphrase @"raw" "" then
+            ByronWalletPutPassphraseData Nothing (ApiT new)
+        else
+            ByronWalletPutPassphraseData (Just $ ApiT old) (ApiT new)
+
+instance PassphraseUpdate WalletPutPassphraseData where
+    producePassphraseData (OldNewPassword old new) =
+        WalletPutPassphraseData (ApiT old) (ApiT new)
```
Comment (KtorZ) [inline lib/cli/src/Cardano/CLI.hs]: The extra parameter and instances above bring quite some overhead. I'd suggest to leave the command as it is. It won't be possible to update legacy wallets that have no or a invalid passphrase set via the command-line. That's an acceptable limitation :+1:

File: lib/core/src/Cardano/Wallet/Api/Types.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: done

File: lib/cli/src/Cardano/CLI.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/cli/src/Cardano/CLI.hs]: reverted support for CLI

**Comment by paweljakubas** [comment]
bors try

File: lib/core/src/Cardano/Wallet/Api/Client.hs [hunk 3]
```diff
@@ -228,7 +229,7 @@ byronWalletClient =
             , listWallets = _listWallets
             , postWallet = _postWallet
             , putWallet = _putWallet
-            , putWalletPassphrase = error "TODO: putWalletPassphrase"
+            , putWalletPassphrase = error "putWalletPassphrase: not impl for byron"
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Client.hs]: This should be `_putByonWalletPassphrase` now, no :) ?

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 4]
```diff
@@ -899,7 +908,11 @@ postRandomWallet ctx body = do
         W.createWallet @(WorkerCtx ctx) @s @k wrk wid wName s
   where
     wName = getApiT (body ^. #name)
-    pwd   = preparePassphrase EncryptWithPBKDF2 $ getApiT (body ^. #passphrase)
+    pwd  = case body ^. #passphrase of
+        Nothing ->
+            preparePassphrase EncryptWithPBKDF2 $ Passphrase ""
+        Just (ApiT (Passphrase passwd)) ->
+            preparePassphrase EncryptWithPBKDF2 $ Passphrase passwd
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: suggestion:

```
pwd = preparePassphrase EncryptWithPBKDF2 $ fromMaybe mempty $ getApiT (body ^. #passphrase)
```

File: lib/core/src/Cardano/Wallet/Api/Client.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Client.hs]: I have this problem of using `ByronWalletPutPassphraseData` and not `WalletPutPassphraseData` in `WalletClient` and it contributed to the solution you did not liked  done in commit https://github.com/input-output-hk/cardano-wallet/pull/1495/commits/e485e7bad1fd04c6fab3b7997715a354b0d60e85

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 5]
```diff
@@ -430,7 +430,7 @@ data SomeByronWalletPostData
 data ByronWalletPostData mw = ByronWalletPostData
     { mnemonicSentence :: !(ApiMnemonicT mw)
     , name :: !(ApiT WalletName)
-    , passphrase :: !(ApiT (Passphrase "raw"))
+    , passphrase :: !(Maybe (ApiT (Passphrase "byron-raw")))
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]: Why this change? The "standard" API shouldn't allow creating Byron wallets with an empty passphrase, this should only be possible through the legacy endpoint when providing the encrypted XPrv. Sorry for the confusion, I misread your question on slack :(, I thought you were talking about the put passphrase endpoint.

Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: ok, then we cannot test it easily. I mean now we can and it passes but I can remove the last two commits and should be fine

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 6]
```diff
@@ -1136,6 +1161,27 @@ putWalletPassphrase ctx (ApiT wid) body = do
...
+putByronWalletPassphrase ctx (ApiT wid) body = do
+    let (ByronWalletPutPassphraseData oldM (ApiT new)) = body
+    withWorkerCtx ctx wid liftE liftE $ \wrk -> liftHandler $
+        case oldM of
+            Just (ApiT (Passphrase old')) -> do
+                let old = Passphrase @"raw" old'
+                W.updateWalletPassphrase wrk wid (old, new)
+            _ -> do
+                let old = Passphrase @"raw" ""
+                W.updateWalletPassphrase wrk wid (old, new)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: suggestion:

```hs
let old = maybe mempty (coerce . getApiT) oldM  
W.updateWalletPassphrase wrk wid (old, new)
```

File: lib/core/src/Cardano/Wallet/Api/Client.hs (hunk 3 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Api/Client.hs]: Okay, I'll look into this :+1:

File: lib/core/src/Cardano/Wallet/Api/Types.hs (hunk 5 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: ~~Hmmm. Okay, let's keep it.~~ .. That's a tricky problem. Because now, we need to change the passphrase type on every endpoint that requires a passphrase.

**Review by KtorZ — APPROVED** [review]

**Comment by KtorZ** [comment]
Another occurrence of the "500" error sometimes happening on CI these days. Merging anyway.
